_posts/2018-12-19-ets-oddity.md

```diff
@@ -0,0 +1,97 @@
+                [set, public, {write_concurrency, true}]).
+#Ref<0.1705802953.985792516.98626>
+> P1 = spawn(fun() ->
+               ets:insert(Tab, {1, a}),
+               ets:insert(Tab, {2, a})
+             end).
+> K1 = ets:first(Tab), K2 = ets:next(Tab, K1).
+```
+
+What are the theoretical possible values of `K1` and `K2`? Let us first list the obvious:
+
+* `K1 = 1`, `K2 = 2` - both values inserted and found in term order
+* `K1 = 2`, `K2 = 1` - since this is a `set`, the hash algorithm may put 2 before 1
+* `K1 = 1`, `K2 = '$end\_of\_table'` - only 1 had time to be inserted
+* `K1 = '$end\_of\_table'`, `K2 = badarg` - no elements were inserted
+
+However it is also possible to get:
+
+* `K1 = 2`, `K2 = '$end\_of\_table'`
+
+This was, at first, very counter-intuitive to me. How can the `ets:first/1` find the
+second value inserted, but then when iterating not find the value inserted before it?
+
+The answer can be found in the way that the `write_concurrency` functionality is
+implemented. Imagine we have a [hash table](https://en.wikipedia.org/wiki/Hash_table)
+where each bucket is protected by a mutex. When inserting a new element the mutex for
+the current bucket has to be taken and when iterating over the hash table we take
+each mutex in turn for the buckets we iterate through.
+
+### Initial Table:
+
+| Bucket #      | Values        |
+| ------------- | ------------- |
+| 1             | `[]`          |
+| 2             | `[]`          |
+| 3             | `[]`          |
+| 4             | `[]`          |
+
+### Finished Table:
+
+| Bucket #      | Values        |
+| ------------- | ------------- |
+| 1             | `[{1,a}]`     |
+| 2             | `[]`          |
+| 3             | `[]`          |
+| 4             | `[{2,b}]`     |
+
+So, in the scenario that leads to the strange behaviour the following will happen:
+
+* ets:first/1 is called when the table is empty and iterates to Bucket #2.
+
+| Bucket #      | Values        |
+| ------------- | ------------- |
+| 1             | `[]`          |
+| 2 (first)     | `[]`          |
+| 3             | `[]`          |
+| 4             | `[]`          |
+
+
+* The OS does a context switch and P1 is allowed to run.
+* P1 inserts both `{1,a}` and `{2,b}` and then exits.
+
+| Bucket #      | Values        |
+| ------------- | ------------- |
+| 1             | `[{1,a}]`     |
+| 2 (first)     | `[]`          |
+| 3             | `[]`          |
+| 4             | `[{2,b}]`     |
+
+* The `ets:first/1` call resumes and will only see `2` and then `'$end\_of\_table'`.
+
+When spelled out like this it becomes more logical that it is possible to get only
+the element inserted as the second element. This is not normally a problem for tables
+of type `set` as you don't normally depend on the order of insertion.
+
+However, for `ordered_set` you may very well depend on the fact that if you see
+`2` you know that you will also have seen any elements inserted before it. But for
```

Comment:
However, for `ordered_set` you may very well depend on the defined iteration order and expect `ets:first` to actually return a key that has a least been first in the table at some point in time.